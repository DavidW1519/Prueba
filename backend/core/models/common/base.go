/*
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to You under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package common

import (
	"encoding/json"
	"github.com/apache/incubator-devlake/core/errors"
	"github.com/spf13/cast"
	"regexp"
	"time"
)

type Model struct {
	ID        uint64    `gorm:"primaryKey" json:"id"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

type ScopeConfig struct {
	Model
	Entities []string `gorm:"type:json;serializer:json" json:"entities" mapstructure:"entities"`
}

type NoPKModel struct {
	CreatedAt     time.Time `json:"createdAt"`
	UpdatedAt     time.Time `json:"updatedAt"`
	RawDataOrigin `swaggerignore:"true"`
}

// embedded fields for tool layer tables
type RawDataOrigin struct {
	// can be used for flushing outdated records from table
	RawDataParams string `gorm:"column:_raw_data_params;type:varchar(255);index" json:"_raw_data_params"`
	RawDataTable  string `gorm:"column:_raw_data_table;type:varchar(255)" json:"_raw_data_table"`
	// can be used for debugging
	RawDataId uint64 `gorm:"column:_raw_data_id" json:"_raw_data_id"`
	// we can store record index into this field, which is helpful for debugging
	RawDataRemark string `gorm:"column:_raw_data_remark" json:"_raw_data_remark"`
}

type GetRawDataOrigin interface {
	GetRawDataOrigin() *RawDataOrigin
}

func (c *RawDataOrigin) GetRawDataOrigin() *RawDataOrigin {
	return c
}

func NewNoPKModel() NoPKModel {
	now := time.Now()
	return NoPKModel{
		CreatedAt: now,
		UpdatedAt: now,
	}
}

var (
	DUPLICATE_REGEX = regexp.MustCompile(`(?i)\bduplicate\b`)
)

func IsDuplicateError(err error) bool {
	return err != nil && DUPLICATE_REGEX.MatchString(err.Error())
}

// CreateRawDataParams TODO in the future use explicit fields instead of a generic params struct, and return a JSON string
// with standardized keys
func CreateRawDataParams(pluginName string, params any) (string, errors.Error) {
	// Append the plugin name to the params struct and serialize to string
	paramsMap := map[string]any{
		"Plugin": pluginName,
	}
	// TODO: maybe sort it to make it consistent
	paramsBytes, err := json.Marshal(params)
	if err != nil {
		return "", errors.Convert(err)
	}
	err = json.Unmarshal(paramsBytes, &paramsMap)
	if err != nil {
		return "", errors.Convert(err)
	}
	// compare all values to strings
	for k, v := range paramsMap {
		paramsMap[k] = cast.ToString(v)
	}
	paramsBytes, err = json.Marshal(paramsMap)
	if err != nil {
		return "", errors.Convert(err)
	}
	return string(paramsBytes), nil
}
