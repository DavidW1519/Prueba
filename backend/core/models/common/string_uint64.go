/*
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to You under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package common

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"github.com/spf13/cast"
	"reflect"
)

type StringUint64 struct {
	v uint64
	t string
}

func NewStringUint64(value uint64) *StringUint64 {
	return &StringUint64{
		v: value,
		t: "uint64",
	}
}

func NewStringUint64FromAny(value interface{}) *StringUint64 {
	return &StringUint64{
		v: cast.ToUint64(value),
		t: "string",
	}
}

func (f *StringUint64) MarshalJSON() ([]byte, error) {
	return json.Marshal(f.String())
}

func (f *StringUint64) Uint64() uint64 {
	if f == nil {
		return 0
	}
	return f.v
}

func (f *StringUint64) String() string {
	//if f.t == "string" {
	//	return fmt.Sprintf("\"%v\"", f.v)
	//}
	return fmt.Sprintf("%v", f.v)
}

func (f *StringUint64) UnmarshalJSON(data []byte) error {
	var i interface{}
	if err := json.Unmarshal(data, &i); err != nil {
		return err
	}
	switch i.(type) {
	case uint64:
		f.t = "uint64"
	case string:
		f.t = "string"
	}
	value, err := cast.ToUint64E(i)
	if err != nil {
		return err
	}
	f.v = value
	return nil
}

func (f *StringUint64) Value() (driver.Value, error) {
	if f == nil {
		return nil, nil
	}
	return f.v, nil
}

func (f *StringUint64) Scan(v interface{}) error {
	switch value := v.(type) {
	case uint8, uint16, uint32, uint64, int8, int16, int32, int64, int, uint:
		*f = StringUint64{
			v: cast.ToUint64(value),
			t: "uint64",
		}
	case string:
		*f = StringUint64{
			v: cast.ToUint64(value),
			t: "string",
		}
	default:
		return fmt.Errorf("[StringUint64] %+v is an unknown type, type: %v with value: %v", v, reflect.TypeOf(v), value)
	}
	return nil
}
