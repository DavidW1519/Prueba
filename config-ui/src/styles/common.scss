.container {
  display: flex;
  justify-content: space-evenly;
  padding-top: 50px;
  min-height: 100vh;
  overflow: hidden;
}

.main {
  padding: 4rem 4rem 4rem calc(250px + 4rem);
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
}

.description {
  display: block;
  margin-bottom: 1.6rem;
  max-width: 600px;
  font-family: 'Montserrat', sans-serif;
}

.logo {
  max-width: 60px;
  margin-bottom: 0.5rem;
}

.code {
  margin-right: 0.2rem;
}

.tagInput {
  position: relative;
  padding-right: 2rem;
}

.input,
.helperInput {
  margin-left: 0.75rem;
}

.input input,
.helperInput input {
  margin-top: 0 !important;
}

.input, .helperInput, .form {
  width: 100%;
}

.helperInput {
  margin-top: -10px;
  width: calc(100% - 12px) !important;
}

.requiredStar {
  color: #F83E3E;
  font-weight: 600;
  font-size: 1.2rem;
  margin-left: 0.25rem;
}

.codeArea {
  width: 100%;
  max-width: 600px !important;
  height: 384px !important;
  font-family: "JetBrains Mono", monospace;
  font-size: 13px;
  border: 0;
  outline: none !important;
  box-shadow: none;
  background-color: #f9f9f9;
}

.formSection {
  margin-top: 10px;
  width: 100%;
  max-width: 600px;
}

.helpIcon {
  margin-left: 0.5rem;
  position: relative;
  top: -0.4rem;
}

.formGroup,
.statusMappingsFilterFormGroup {
  width: 100%;
  max-width: 600px;
}

.formGroup div,
.statusMappingsFilterFormGroup div {
  color: #717484 !important;
}

.formGroup label,
.statusMappingsFilterFormGroup label {
  color: #4B4D58;
  display: flex;
  align-items: center;
  margin-right: 0 !important;
}

.button {
  margin-left: 1rem;
}

.formContainer,
.jiraFormContainer {
  display: flex;
  width: 100%;
  align-items: center;
  margin-bottom: 0.4rem;
}

.jiraFormContainer {
  height: auto;
  flex-wrap: wrap;
  flex-direction: column;
  align-content: flex-start;
}

.jiraFormContainerItem {
  width: 100%;
}

.statusMappingsFilterFormGroup {
  align-self: flex-start;
}

.mappingTagStatus {
  box-sizing: border-box;
}

.headlineContainer {
  width: 100%;
}

.headlineContainer h1 {
  margin: 0 0 0.5rem;
}

.headline {
  margin-top: 1.5rem;
  margin-bottom: 0.25rem;
}

.saveBtn {
  margin: 2rem auto 0 0;
  display: block;
  background-color: #717484 !important;
  color: #fff !important;
}

.saveBtn:hover {
  opacity: 0.9;
}

.pluginSpinner {
  display: inline-block;
  margin-right: 0.25rem;
  margin-top: -0.25rem;
}

.pluginSpinnerWrap {
  margin-top: 0.5rem;
}

.statusTabs {
  width: 100%;
}

.addNewStatusBtn {
  align-self: flex-start;
}

.addNewStatusBtn,
.addNewStatusBtnDialog {
  border-radius: 3px;
  border: 1px solid rgba(0,0,0,0.15);
  padding: 0.5rem 0.5rem;
  outline: 0;
  display: inline-block;
  margin-bottom: 0.75rem;
}

.addNewStatusBtnDialog {
  margin-top: 1rem;
  text-align: center;
  width: 100%;
}

.customStatusFormGroup {
  margin-bottom: -0.5rem;
  padding: 2rem 3rem;
}

.customStatusDialog {
  background-color: #F4F4F6;
  width: 100%;
  max-width: 664px;
  height: auto;
}
