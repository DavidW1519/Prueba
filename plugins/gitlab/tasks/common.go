package tasks

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/merico-dev/lake/config"
	"github.com/merico-dev/lake/plugins/core"
)

type GitlabApiClient struct {
	core.ApiClient
}

var gitlabApiClient *GitlabApiClient

func CreateApiClient() *GitlabApiClient {
	if gitlabApiClient == nil {
		gitlabApiClient = &GitlabApiClient{}
		gitlabApiClient.Setup(
			config.V.GetString("GITLAB_ENDPOINT"),
			map[string]string{
				"Authorization": fmt.Sprintf("Bearer %v", config.V.GetString("GITLAB_AUTH")),
			},
			10*time.Second,
			3,
		)
	}
	return gitlabApiClient
}

type GitlabPaginationHandler func(res *http.Response) error

func (gitlabApiClient *GitlabApiClient) FetchWithPagination(resourceUri string, page string, pageSize string, handler GitlabPaginationHandler) error {

	var resourceUriFormat string
	if strings.ContainsAny(resourceUri, "?") {
		resourceUriFormat = resourceUri + "&per_page=%v&page=%v"
	} else {
		resourceUriFormat = resourceUri + "?per_page=%v&page=%v"
	}

	for page != "" {
		res, err := gitlabApiClient.Get(fmt.Sprintf(resourceUriFormat, pageSize, page), nil, nil)

		if err != nil {
			return err
		}
		page = res.Header.Get("x-next-page")
		handler(res)
	}
	return nil
}
