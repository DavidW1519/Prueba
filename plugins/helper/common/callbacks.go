package common

import "net/http"

type ApiAsyncCallback func(*http.Response) error
