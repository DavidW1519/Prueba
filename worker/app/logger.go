package app

import (
	"fmt"

	"github.com/apache/incubator-devlake/plugins/core"
	"go.temporal.io/sdk/log"
)

type TemporalLogger struct {
	log core.Logger
}

func NewTemporalLogger(log core.Logger) log.Logger {
	return &TemporalLogger{
		log,
	}
}

func (l *TemporalLogger) Log(lv core.LogLevel, msg string, keyvals ...interface{}) {
	if l.log.IsLevelEnabled(lv) {
		for i := 0; i < len(keyvals); i += 2 {
			msg += fmt.Sprintf(" %s %v", keyvals[i], keyvals[i+1])
		}
		l.log.Log(lv, msg)
	}
}

func (l *TemporalLogger) Debug(msg string, keyvals ...interface{}) {
	l.Log(core.LOG_DEBUG, msg, keyvals...)
}

func (l *TemporalLogger) Info(msg string, keyvals ...interface{}) {

	l.Log(core.LOG_INFO, msg, keyvals...)
}

func (l *TemporalLogger) Warn(msg string, keyvals ...interface{}) {

	l.Log(core.LOG_WARN, msg, keyvals...)
}

func (l *TemporalLogger) Error(msg string, keyvals ...interface{}) {
	l.Log(core.LOG_ERROR, msg, keyvals...)
}
