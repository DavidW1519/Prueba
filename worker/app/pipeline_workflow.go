/*
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to You under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package app

import (
	"fmt"
	"strings"
	"time"

	"github.com/apache/incubator-devlake/runner"
	"go.temporal.io/sdk/workflow"
)

// DevLakePipelineWorkflow FIXME ...
func DevLakePipelineWorkflow(ctx workflow.Context, configJson []byte, pipelineId uint64) error {
	cfg, logger, db, err := loadResources(configJson)
	if err != nil {
		return err
	}
	logger.Info("received pipeline #%d", pipelineId)
	err = runner.RunPipeline(
		cfg,
		logger,
		db,
		pipelineId,
		func(taskIds []uint64) error {
			futures := make([]workflow.Future, len(taskIds))
			for i, taskId := range taskIds {
				activityOpts := workflow.ActivityOptions{
					ActivityID:          fmt.Sprintf("task #%d", taskId),
					StartToCloseTimeout: 24 * time.Hour,
					WaitForCancellation: true,
				}
				activityCtx := workflow.WithActivityOptions(ctx, activityOpts)
				futures[i] = workflow.ExecuteActivity(activityCtx, DevLakeTaskActivity, configJson, taskId)
			}
			errs := make([]string, 0)
			for _, future := range futures {
				err := future.Get(ctx, nil)
				if err != nil {
					errs = append(errs, err.Error())
				}
			}
			if len(errs) > 0 {
				return fmt.Errorf(strings.Join(errs, "\n"))
			}
			return nil
		},
	)
	if err != nil {
		logger.Error("failed to execute pipeline #%d: %w", pipelineId, err)
	}
	logger.Info("finished pipeline #%d", pipelineId)
	return err
}
