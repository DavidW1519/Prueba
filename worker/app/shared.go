package app

import (
	"bytes"

	"github.com/apache/incubator-devlake/logger"
	"github.com/apache/incubator-devlake/plugins/core"
	"github.com/apache/incubator-devlake/runner"
	"github.com/spf13/viper"
	"gorm.io/gorm"
)

func loadResources(configJson []byte) (*viper.Viper, core.Logger, *gorm.DB, error) {
	// prepare
	cfg := viper.New()
	cfg.SetConfigType("json")
	err := cfg.ReadConfig(bytes.NewBuffer(configJson))
	if err != nil {
		return nil, nil, nil, err
	}
	// TODO: should be redirected to server
	logger := logger.Global.Nested("worker")
	db, err := runner.NewGormDb(cfg, logger)
	if err != nil {
		return nil, nil, nil, err
	}
	return cfg, logger, db, err
}
