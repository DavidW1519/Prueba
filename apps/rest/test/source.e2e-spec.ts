import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';

describe('SourceController (e2e)', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe());
    await app.init();
  });

  describe('/source (POST)', () => {
    it('should return source type', () => {
      const newSource = {
        type: 'jira',
        options: {
          host: 'https://www.atlassian.com/',
          email: 'xx@example.com',
          auth: 'base64EncodedAuthToken',
        },
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(newSource)
        .expect(201)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should return validate error', () => {
      const sourceWithoutType = {
        options: {},
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should return validate error 2', () => {
      const sourceWithoutType = {
        type: 'github',
        options: {},
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should return validate error 3', () => {
      const sourceWithoutType = {
        type: 'gitlab',
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
