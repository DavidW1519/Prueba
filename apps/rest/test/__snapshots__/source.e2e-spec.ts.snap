// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceController (e2e) /source (POST) should return validate error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "type must be one of the following values: jira, gitlab",
  ],
  "statusCode": 400,
}
`;

exports[`SourceController (e2e) /source (POST) should return validate error 2 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "type must be one of the following values: jira, gitlab",
  ],
  "statusCode": 400,
}
`;

exports[`SourceController (e2e) /source (POST) should return validate error 3 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "options should not be empty",
  ],
  "statusCode": 400,
}
`;
